src/patterns.jl

```diff
@@ -0,0 +1,131 @@
+"""
+    AbstractPattern
+
+Abstract supertype of all sparsity pattern representations.
+
+## Type hierarchy
+```
+AbstractPattern
+├── AbstractVectorPattern: used in GradientTracer, ConnectivityTracer
+│   └── IndexSetVectorPattern
+└── AbstractHessianPattern: used in HessianTracer
```

Comment:
Why not `AbstractMatrixVectorPattern`? The naming would be more coherent

Reply:
Ideally, I would have called `AbstractVectorPattern` `AbstractGradientPattern`, but there is the permanently annoying `ConnectivityTracer`.

Do you prefer `AbstractFirstOrderPattern` and `AbstractSecondOrderPattern`?
I'm just a bit afraid of an explosion in subtype-name-length, e.g. `IndexSetSecondOrderPattern`.

Maybe there is a shorter alternative to `FirstOrder` and `SecondOrder`?

Reply:
I renamed it to `AbstractGradientPattern` in anticipation of removing `ConnectivityTracer` in a follow-up PR.